test/tid.c

```diff
@@ -569,197 +569,299 @@ test_id_type_list(void)
             ERROR("invalid obj_list");
-            goto out;
-        } /* end if */
-        else {
-            /* Remove the object.  Mark as "freeing" so its own callback does
-             * not free another object. */
-            obj->obj_list->list[i].freeing = TRUE;
-            ret                            = H5Idec_ref(obj->obj_list->list[i].id);
-            CHECK(ret, FAIL, "H5Idec_ref");
-            if (ret == FAIL)
-                goto out;
-            obj->obj_list->list[i].freeing = FALSE;
-        } /* end else */
-    }     /* end if */
-
-    /* Verify nobjs_rem is non-negative */
-    if (obj->obj_list->nobjs_rem < 0) {
-        ERROR("invalid nobjs_rem");
-        goto out;
-    } /* end if */
+            goto error;
+        }
+
+        /* Mark the object we're about to free so its own callback does
+         * not free another object. We don't want to recursively free the
+         * entire list when we free the first ID.
+         */
+        obj->list->objects[i].freeing = TRUE;
+
+        /* Decrement the reference count on the object */
+        ret = H5Idec_ref(obj->list->objects[i].id);
+        CHECK(ret, FAIL, "H5Idec_ref");
+        if (ret == FAIL)
+            goto error;
+
+        /* Unset the "freeing" flag */
+        obj->list->objects[i].freeing = FALSE;
+    }
+
+    /* Verify the number of objects remaining in the master list is non-negative */
+    if (obj->list->remaining < 0) {
+        ERROR("invalid number of objects remaining");
+        goto error;
+    }
 
     return 0;
 
-out:
+error:
     return -1;
-} /* end test_rct_free() */
+} /* end rct_free_cb() */
 
 /* Test function */
 static int
 test_remove_clear_type(void)
 {
     H5I_type_t      obj_type;
-    test_rct_list_t obj_list;
-    test_rct_obj_t  list[TEST_RCT_MAX_NOBJS];
+    rct_obj_list_t  obj_list;
+    rct_obj_t *     objects = NULL; /* Convenience pointer to objects stored in master list */
+    size_t          list_size;
     long            i, j;
-    long            nobjs_found;
-    hsize_t         nmembers;
     herr_t          ret; /* return value */
 
-    /* Register type */
-    obj_type = H5Iregister_type((size_t)8, 0, test_rct_free);
+    /* Register a user-defined type with our custom ID-deleting callback */
+    obj_type = H5Iregister_type((size_t)8, 0, rct_free_cb);
     CHECK(obj_type, H5I_BADID, "H5Iregister_type");
     if (obj_type == H5I_BADID)
-        goto out;
+        goto error;
+
+    /* Create an array to hold the objects in the master list */
+    list_size = RCT_MAX_NOBJS * sizeof(rct_obj_t);
+    obj_list.objects = HDmalloc(list_size);
+    CHECK(obj_list.objects, NULL, "HDcalloc");
+    if (NULL == obj_list.objects)
+        goto error;
+
+    /* Set a convenience pointer to the object array */
+    objects = obj_list.objects;
+
+    for (i = 0; i < RCT_NITER; i++) {
+
+        /* The number of members in the type, according to the HDF5 library */
+        hsize_t nmembers = 1234567;
```

Comment:
Initializing nmembers to 1234567 here was pretty confusing for me, until I realized that it was just an invalid value for the later call to H5Inmembers().   I would suggest moving the initialization to just before calling H5Inmembers(), to help the next person.